diff --git a/imx/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc b/imx/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc
index afe85a2..67c0ace 100644
--- a/imx/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc
+++ b/imx/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc
@@ -1,12 +1,12 @@
 # Use i.MX Kernel, U-Boot and Gstreamer 1.0 providers
-PREFERRED_PROVIDER_u-boot_mx5 = "u-boot-fslc"
-PREFERRED_PROVIDER_u-boot_mx6 = "u-boot-imx"
-PREFERRED_PROVIDER_u-boot_mx6ul = "u-boot-imx"
-PREFERRED_PROVIDER_u-boot_mx7 = "u-boot-imx"
+#PREFERRED_PROVIDER_u-boot_mx5 = "u-boot-fslc"
+#PREFERRED_PROVIDER_u-boot_mx6 = "u-boot-imx"
+#PREFERRED_PROVIDER_u-boot_mx6ul = "u-boot-imx"
+#PREFERRED_PROVIDER_u-boot_mx7 = "u-boot-imx"
 
-PREFERRED_PROVIDER_virtual/kernel_mx6 = "linux-imx"
-PREFERRED_PROVIDER_virtual/kernel_mx6ul = "linux-imx"
-PREFERRED_PROVIDER_virtual/kernel_mx7 = "linux-imx"
+#PREFERRED_PROVIDER_virtual/kernel_mx6 = "linux-imx"
+#PREFERRED_PROVIDER_virtual/kernel_mx6ul = "linux-imx"
+#PREFERRED_PROVIDER_virtual/kernel_mx7 = "linux-imx"
 
 # Masking out community conflicts
 # Use new Wayland version
