diff --git a/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc b/meta-sdk/conf/distro/include$
index 3bdbaf5bc..b95abaafc 100644
--- a/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc
+++ b/meta-sdk/conf/distro/include/fsl-imx-preferred-env.inc
@@ -1,5 +1,5 @@
-PREFERRED_PROVIDER_virtual/bootloader:imx-nxp-bsp = "u-boot-imx"
-PREFERRED_PROVIDER_virtual/kernel:imx-nxp-bsp = "linux-imx"
+#PREFERRED_PROVIDER_virtual/bootloader:imx-nxp-bsp = "u-boot-imx"
+#PREFERRED_PROVIDER_virtual/kernel:imx-nxp-bsp = "linux-imx"
 PREFERRED_PROVIDER_linux-imx-mfgtool = "linux-imx-mfgtool"

 PREFERRED_PROVIDER_opencl-headers:imxgpu = "imx-gpu-viv
