diff --git a/imx/meta-bsp/recipes-connectivity/connman/connman_%.bbappend b/imx/meta-bsp/recipes-connectivity/connman/connman_%.bbappend
index 64bebb2..7687e6a 100644
--- a/imx/meta-bsp/recipes-connectivity/connman/connman_%.bbappend
+++ b/imx/meta-bsp/recipes-connectivity/connman/connman_%.bbappend
@@ -27,7 +27,7 @@ pkg_postinst_${PN} () {
         if [ -n "$D" ]; then
             OPTS="--root=$D"
         fi
-        systemctl $OPTS mask connman.service
+#        systemctl $OPTS mask connman.service
     fi
 }
 
