diff --git a/meta-bsp/recipes-bsp/imx-mkimage/imx-boot_1.0.bb b/meta-bsp/recipes-bsp/imx-mkimage/imx-boot_1.0.bb
index 9987de761f..947520e9b7 100644
--- a/meta-bsp/recipes-bsp/imx-mkimage/imx-boot_1.0.bb
+++ b/meta-bsp/recipes-bsp/imx-mkimage/imx-boot_1.0.bb
@@ -79,7 +79,8 @@ REV_OPTION:mx8ulp-generic-bsp = \
     "${@bb.utils.contains('MACHINE_FEATURES', 'soc-reva0', '', \
                                                            'REV=A2', d)}"
 REV_OPTION:mx91p-generic-bsp = "REV=A1"
-REV_OPTION:mx93-generic-bsp  = "REV=A1"
+#REV_OPTION:mx93-generic-bsp  = "REV=A1"
+REV_OPTION:mx93-generic-bsp  = "REV=A0"

 compile_mx8m() {
     bbnote 8MQ/8MM/8MN/8MP boot binary build
