diff --git a/recipes-connectivity/murata-binaries/murata-binaries_1.0.bb b/recipes-connectivity/murata-binaries/murata-binaries_1.0.bb
index dc1c9ca..119b9e8 100644
--- a/recipes-connectivity/murata-binaries/murata-binaries_1.0.bb
+++ b/recipes-connectivity/murata-binaries/murata-binaries_1.0.bb
@@ -8,6 +8,7 @@ SRC_URI = " \
 	git://github.com/murata-wireless/cyw-fmac-nvram;protocol=http;branch=mothra;destsuffix=cyw-fmac-nvram \
 	git://github.com/murata-wireless/cyw-bt-patch;protocol=http;branch=rocko-mothra;destsuffix=cyw-bt-patch \
 	git://github.com/murata-wireless/cyw-fmac-utils-imx32;protocol=http;branch=mothra;destsuffix=cyw-fmac-utils-imx32 \
+	git://github.com/murata-wireless/cyw-fmac-utils-imx64;protocol=http;branch=mothra;destsuffix=cyw-fmac-utils-imx64 \
 	file://10-network.rules \
 "
 
@@ -15,6 +16,7 @@ SRCREV_cyw-fmac-fw="d810709364dce9771394e787f3366cac8ef82020"
 SRCREV_cyw-fmac-nvram="7946555202d33556639821ec5c15b2b76bad2e6d"
 SRCREV_cyw-bt-patch="80995c1d8bf1546dbc7be77fef1cd8f616675527"
 SRCREV_cyw-fmac-utils-imx32="95b79c0a2064e742f3605c3937d93f1c0a377587"
+SRCREV_cyw-fmac-utils-imx64="a9a053cb6ca8778739297f152dd194474f0068f7"
 
 SRCREV_default = "${AUTOREV}"
 
@@ -40,6 +42,10 @@ do_compile () {
 PACKAGES_prepend = "murata-binaries-wlarm "
 FILES_murata-binaries-wlarm = "${bindir}/wlarm"
 
+DO_INSTALL_64BIT_BINARIES = "no"
+DO_INSTALL_64BIT_BINARIES_mx6 = "no"
+DO_INSTALL_64BIT_BINARIES_mx7 = "no"
+DO_INSTALL_64BIT_BINARIES_mx8 = "yes"
 
 do_install () {
 	echo "Installing: "
@@ -97,7 +103,11 @@ do_install () {
 	install -m 444 ${S}/10-network.rules                  ${D}${sysconfdir}/udev/rules.d/10-network.rules
 
 #       Copying wl tool binary to /usr/sbin
-	install -m 755 ${S}/cyw-fmac-utils-imx32/wl ${D}/usr/sbin/wl
+	if [ ${DO_INSTALL_64BIT_BINARIES} = "yes" ]; then
+		install -m 755 ${S}/cyw-fmac-utils-imx64/wl ${D}/usr/sbin/wl
+	else
+		install -m 755 ${S}/cyw-fmac-utils-imx32/wl ${D}/usr/sbin/wl
+	fi
 }
 
 PACKAGES =+ "${PN}-mfgtest"
